### PR TITLE
dgud/mnesia/type specs

### PR DESCRIPTION
**Pull request by dgud**
Add typespecs to mnesia api.

### CODE REVIEW

File: lib/mnesia/src/mnesia.erl [hunk 1]
```diff
@@ -363,28 +404,60 @@ non_transaction(State, Fun, Args, ActivityKind, Mod)
 non_transaction(_State, Fun, Args, _ActivityKind, _Mod) ->
     {aborted, {badarg, Fun, Args}}.
 
+-spec async_dirty(Fun) -> Res | no_return() when
```
Comment (kostis) [inline lib/mnesia/src/mnesia.erl]: It's a common misconception, but the type `foo() | no_return()` does not make much sense.  For starters, it's equivalent to just `foo()`.  Second, _all_ non-total functions in Erlang have an implicit no_return(), for example to handle all cases that they crash (or loop) for some reason.  I suggest you take out `| no_return()`.

File: lib/mnesia/src/mnesia.erl [hunk 2]
```diff
@@ -138,6 +138,21 @@
 -include("mnesia.hrl").
 -import(mnesia_lib, [verbose/2]).
 
+-type t_result(Res) :: {'atomic', Res} | {'aborted', Reason::term()}.
+-type activity() :: 'ets' | 'async_dirty' | 'sync_dirty' | 'transaction' | 'sync_transaction' |
+                    {'transaction', Retries::non_neg_integer()} |
+                    {'sync_transaction', Retries::non_neg_integer()}.
+-type table() :: 'atom'.
```
Comment (kostis) [inline lib/mnesia/src/mnesia.erl]: Is this a typo?  Should it say `atom()` instead?

File: lib/mnesia/src/mnesia.erl [hunk 3]
```diff
@@ -138,6 +138,21 @@
...
+-type storage_type() :: 'ram_copies' | 'disc_copies' | 'disc_only_copies'.
+-type index_attr() :: 'atom' | non_neg_integer().
```
Comment (kostis) [inline lib/mnesia/src/mnesia.erl]: Similar comment as above...

File: lib/mnesia/src/mnesia.erl [hunk 4]
```diff
@@ -1614,23 +1752,28 @@ do_dirty_update_counter(_SyncMode, Tab, _Key, Incr) ->
...
 
+-spec dirty_match_object(Pattern::tuple()) -> [Record::tuple()].
 dirty_match_object(Pat) when is_tuple(Pat), tuple_size(Pat) > 2 ->
     Tab = element(1, Pat),
     dirty_match_object(Tab, Pat);
 dirty_match_object(Pat) ->
     abort({bad_type, Pat}).
 
+-spec dirty_match_object(Tab,Pattern) -> [Record] when
+      Tab::table(),Pattern::tuple(),Record::tuple().
```
Comment (kostis) [inline lib/mnesia/src/mnesia.erl]: Shouldn't there be spaces after the commas here?

File: lib/mnesia/src/mnesia.erl [hunk 5]
```diff
@@ -2361,96 +2531,180 @@ load_mnesia_or_abort() ->
...
 backup_checkpoint(Name, Opaque) ->
     mnesia_log:backup_checkpoint(Name, Opaque).
 
+-spec backup_checkpoint(Name::_, Dest::_, Mod::module()) ->
+                               'ok' | {'error', Reason::term()}.
 backup_checkpoint(Name, Opaque, Mod) ->
     mnesia_log:backup_checkpoint(Name, Opaque, Mod).
 
+-spec restore(Src::_, [Arg]) -> t_result([table()]) when
+      Op  :: 'skip_tables' | 'clear_tables' | 'keep_tables' | 'restore_tables',
```
Comment (kostis) [inline lib/mnesia/src/mnesia.erl]: I suggest you declare an appropriate type (`op()` or similar) for this union.

File: lib/mnesia/src/mnesia.erl [hunk 6]
```diff
@@ -2361,96 +2531,180 @@ load_mnesia_or_abort() ->
...
+      Arg :: {'module', module()} | {Op, [table()]} | {'default_op', Op}.
 restore(Opaque, Args) ->
     mnesia_schema:restore(Opaque, Args).
 
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 %% Table mgt
 
+-spec create_table([Arg]) -> t_result('ok') when
+      Arg :: {'name', table()} |
```
Comment (kostis) [inline lib/mnesia/src/mnesia.erl]: There is a lot of redundancy here, for essentially the same functions.  I suggest you introduce a type for the common parts of the Arguments.  Also, perhaps the `'set' | 'ordered_set' | 'bag'` part can be a remote type from `ets` (`duplicate_bag` is not allowed here?).

File: lib/mnesia/src/mnesia.erl [hunk 7]
```diff
@@ -2610,31 +2879,39 @@ set_master_nodes(Tab, Nodes) ->
...
+-spec dump_log() -> 'dumped'.
 dump_log() ->
     mnesia_controller:sync_dump_log(user).
 
+-spec sync_log() -> 'ok' | {'error', Reason::term()}.
 sync_log() ->
     mnesia_monitor:sync_log(latest_log).
 
+-spec subscribe(What) -> {'ok', node()} | {'error', Reason::term()} when
+      What :: 'system' | 'activity' | {'table', table(), 'simple' | 'detailed'}.
```
Comment (kostis) [inline lib/mnesia/src/mnesia.erl]: `'simple' | 'detailed'` should be a type, both so that it has a descriptive name and avoid code duplication.

**Comment by kostis** [comment]
Ideally, these two commits should have been squashed.

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

File: lib/mnesia/src/mnesia.erl (hunk 1 above)
Comment (dgud) [inline lib/mnesia/src/mnesia.erl]: I know it have no meaning for dialyzer,  but I believe it does for documentation,
the transaction functions catch errors and return {aborted, Reason}

**Comment by dgud** [comment]
> Ideally, these two commits should have been squashed.

That will be done eventually, and is what fixup! commit does in git rebase.

File: lib/mnesia/src/mnesia.erl (hunk 6 above)
Comment (dgud) [inline lib/mnesia/src/mnesia.erl]: duplicate_bag is not allowed in mnesia.

**Comment by dgud** [comment]
Merged

**Pull request closed without merging**